test_cases/stdlib/check_functools.py

```diff
@@ -0,0 +1,28 @@
+from __future__ import annotations
+
+import sys
+
+if sys.version_info >= (3, 8):
+    from functools import cached_property
+    from typing import cast
+    from typing_extensions import assert_type
+
+    class A:
+        def __init__(self, x: int):
+            self.x = x
+
+        @cached_property
+        def x(self) -> int:
+            return 0
+
+    assert_type(A(x=1).x, int)
+
+    class B:
+        @cached_property
+        def x(self) -> int:
+            return 0
+
+    b = B()
+    assert_type(b.x, int)
+    b.x = cast(int, 4)
```

Comment:
Heh, is the `cast()` because pyright infers the type as being `Literal[4]` otherwise? If so, I prefer this little workaround:

```python
def check_cached_property_settable(x: int) -> None:
    b = B()
    assert_type(b.x, int)
    b.x = x
    assert_type(b.x, int)
```

Reply:
Actually, you might even be able to get away with just this:

```python
b = B()
assert_type(b.x, int)
x: int = 4
b.x = x
assert_type(b.x, int)
```

Reply:
The second one didn't work -- pyright was still narrowing the type. I went with the first one.